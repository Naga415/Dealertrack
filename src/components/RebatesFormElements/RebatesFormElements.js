import React from "react";
import PropTypes from "prop-types";
import Input from "../customizedUIComponents/Input/Input";
import Row from "@coxautokc/fusion-ui-components/lib/Row";
import Col from "@coxautokc/fusion-ui-components/lib/Col";
import "./RebatesFormElements.scss";

export default class RebatesFormElements extends React.Component {
  generateOnChangeHandler = (r, type, position) => {
    if (type !== "DealerRebate") {
      return e => {
        this.props.onChange(type, e.target.value);
      };
    }

    return e => {
      this.props.onChange(type, e.target.value, position);
    };
  };
  generateOnBlurHandler = (r, type, position) => {
    if (type !== "DealerRebate") {
      return e => {
        this.props.onBlur(type, e.target.value);
      };
    }

    return e => {
      this.props.onBlur(type, e.target.value, position);
    };
  };
  generateOnDealerRebateRemoveHandler = position => {
    return () => {
      this.props.onDealerRebateRemove(position);
    };
  };
  renderRebateField(r, type, position) {
    return (
      <Row key={`rebates-row-${type}${position ? `-${position}` : ""}`}>
        {type !== "DealerRebate" ? (
          <Col md={4} xs={4}>
            <label htmlFor={r.type}>{r.name}</label>
          </Col>
        ) : (
          ""
        )}
        <Col
          xs={type !== "DealerRebate" ? 5 : 12}
          md={type !== "DealerRebate" ? 5 : 12}
        >
          <Input
            htmlId={type + (position ? ('-' + position + '-') : '-') + 'Input'}
            name={type + (position ? ('-' + position + '-') : '-') + 'Input'}
            disabled={type !== "DealerRebate"}
            addOnIndentation="left"
            addOnClassName="fa fa-usd"
            value={r.amount}
            placeholder="0.00"
            onChange={this.generateOnChangeHandler(r, type, position)}
            onBlur={this.generateOnBlurHandler(r, type, position)}
            error={this.props.showErrors && r.error}
          />
        </Col>
      </Row>
    );
  }
  render() {
    const { rebates, onDealerRebateAdd, showErrors } = this.props;
    const nonDealerRebateFields = rebates.filter(r => r.type !== "DealerRebate");
    const dealerRebateFields = rebates.find(r => r.type === "DealerRebate")
      .subElements;
    const maxDealerRebateCountReached = dealerRebateFields.length >= 4;
    const minDealerRebateCountReached = dealerRebateFields.length <= 1;

    return (
      <div id="RebatesFormElements">
        {nonDealerRebateFields.map(r => this.renderRebateField(r, r.type))}
        <Row>
          <Col md={4} xs={4}>
            <label htmlFor="dealerRebate">Dealer Rebate</label>
          </Col>
          <Col md={5} xs={5}>
            {dealerRebateFields.map((r, index) =>
              this.renderRebateField(r, "DealerRebate", index + 1)
            )}
          </Col>
          <Col md={3} xs={3}>
            {dealerRebateFields.map((r, index) => (
              <Row key={index} className="extra-padded-row">
                <Col xs={6} md={6} className="dealer-rebate-add">
                  <a 
                    id={"dealer-rebate-add" + "-" + (index + 1)}
                    className={(maxDealerRebateCountReached || ((index + 1) !== dealerRebateFields.length)) ? 'disabled': ''} 
                    disabled={maxDealerRebateCountReached || ((index + 1) !== dealerRebateFields.length)}
                    onClick={!(maxDealerRebateCountReached || ((index + 1) !== dealerRebateFields.length)) && onDealerRebateAdd}>
                    <i className="fa fa-plus"/>
                  </a>
                </Col>
                <Col xs={6} md={6} className="dealer-rebate-remove">
                  <a
                    id={"dealer-rebate-remove" + "-" + (index + 1)}
                    className={minDealerRebateCountReached ? 'disabled': ''} 
                    disabled={minDealerRebateCountReached}
                    onClick={
                      !minDealerRebateCountReached &&
                      this.generateOnDealerRebateRemoveHandler(index + 1)
                    }>
                    <i className="fa fa-minus"/>
                  </a>
                </Col>
              </Row>
            ))}           
          </Col>
        </Row>
        {showErrors && (
          <Row>
            <Col xs={12} md={12} className="error">
              * The value must be from $0 to $99,999.99
            </Col>
          </Row>
        )}
      </div>
    );
  }
}

RebatesFormElements.propTypes = {
  showErrors: PropTypes.bool,
  rebates: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
  onDealerRebateAdd: PropTypes.func.isRequired,
  onDealerRebateRemove: PropTypes.func.isRequired,
};