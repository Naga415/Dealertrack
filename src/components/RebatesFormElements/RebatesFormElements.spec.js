"use strict";
import React from "react";
import sinon from "sinon";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import cloneDeep from "lodash/cloneDeep";
import RebatesFormElements from "./RebatesFormElements";
import 'jsdom-global/register';

describe("RebatesFormElements specs", () => {
  const rebates = [
    {
      id: 1,
      type: "Incentives",
      name: "Incentives",
      amount: "6465"
    },
    {
      id: 2,
      type: "LoyaltyCash",
      name: "Loyalty Cash",
      amount: "565"
    },
    {
      type: "DealerRebate",
      name: "Dealer Rebate",
      subElements: [
        {
          id: 3,
          amount: "5454"
        },
        {
          id: 4,
          amount: "321"
        },
        {
          id: 5,
          amount: "33"
        }
      ]
    }
  ];
    
  const handleOnInputChange = sinon.spy();
  const handleOnInputBlur = sinon.spy();
  const handleOnAddDealerAmount = sinon.spy();
  const handleOnRemoveDealerAmount = sinon.spy();

  let component, props;

  beforeEach((done) => {
    props = {
      rebates,
      onChange: handleOnInputChange,
      onBlur: handleOnInputBlur,
      onDealerRebateAdd: handleOnAddDealerAmount,
      onDealerRebateRemove: handleOnRemoveDealerAmount
    };
    component = mount(<RebatesFormElements {...props} />);
    done();
  });

  afterEach((done) => {
    handleOnInputChange.reset();
    handleOnInputBlur.reset();
    handleOnAddDealerAmount.reset();
    handleOnRemoveDealerAmount.reset();
    done();
  });

  it("should render <RebatesFormElements />", () => {
    component = renderer.create(<RebatesFormElements {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should call onChange handler when any input changes", () => {
    component.find("#Incentives-Input").simulate("change", { target: { value: "1"}});
    component.find("#LoyaltyCash-Input").simulate("change", { target: { value: "1"}});
    component.find("#DealerRebate-1-Input").simulate("change", { target: { value: "1"}});
    component.find("#DealerRebate-2-Input").simulate("change", { target: { value: "1"}});
    component.find("#DealerRebate-3-Input").simulate("change", { target: { value: "1"}});
    expect(handleOnInputChange.callCount).toEqual(5);
  });

  it("should call onChange handler when any input blurs out", () => {
    component.find("#Incentives-Input").simulate("blur");
    component.find("#LoyaltyCash-Input").simulate("blur");
    component.find("#DealerRebate-1-Input").simulate("blur");
    component.find("#DealerRebate-2-Input").simulate("blur");
    component.find("#DealerRebate-3-Input").simulate("blur");
    expect(handleOnInputBlur.callCount).toEqual(5);
  });

  it("should disable remove icon for a dealer rebate when minimum limit is reached", () => {
    const updatedProps = cloneDeep(props);
    updatedProps.rebates[2].subElements = [{
      id: 6,
      amount: "44"
    }]; // Minimum limit is reached
    component = mount(<RebatesFormElements {...updatedProps} />);
    expect(component.find("#dealer-rebate-remove-1.disabled").length).toEqual(1);
  });


  it("should disable add icon for all dealer rebate except the last one when maximum limit is not reached", () => {
    expect(component.find("#dealer-rebate-add-1.disabled").length).toEqual(1);
    expect(component.find("#dealer-rebate-add-2.disabled").length).toEqual(1);
    expect(component.find("#dealer-rebate-add-3.disabled").length).toEqual(0);
  });


  it("should disable add icon for all dealer rebates when maximum limit is reached", () => {
    const updatedProps = cloneDeep(props);
    updatedProps.rebates[2].subElements.push({
      id: 6,
      amount: "44"
    }); // Maximum limit is reached
    component = mount(<RebatesFormElements {...updatedProps} />);
    expect(component.find("#dealer-rebate-add-1.disabled").length).toEqual(1);
    expect(component.find("#dealer-rebate-add-2.disabled").length).toEqual(1);
    expect(component.find("#dealer-rebate-add-3.disabled").length).toEqual(1);
    expect(component.find("#dealer-rebate-add-4.disabled").length).toEqual(1);
  });

  it("should call onDealerAdd handler when maximum limit is not reached and add icon is clicked on the last delaer rebate", () => {
    component.find("#dealer-rebate-add-3").simulate("click"); // click on last one
    expect(handleOnAddDealerAmount.calledOnce).toEqual(true);
  });

  it("should call onDealerRemove hanlder when minimum limit is not reached and remove icon is clicked on any of the dealer rebates", () => {
    component.find("#dealer-rebate-remove-1").simulate("click"); // click on last one
    expect(handleOnRemoveDealerAmount.calledWith(1)).toEqual(true);
    component.find("#dealer-rebate-remove-2").simulate("click"); // click on last one
    expect(handleOnRemoveDealerAmount.calledWith(2)).toEqual(true);
    component.find("#dealer-rebate-remove-3").simulate("click"); // click on last one
    expect(handleOnRemoveDealerAmount.calledWith(3)).toEqual(true);
  });
});
