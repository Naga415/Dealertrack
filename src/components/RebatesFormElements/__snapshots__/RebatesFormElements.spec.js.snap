// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RebatesFormElements specs should render <RebatesFormElements /> 1`] = `
<div
  id="RebatesFormElements"
>
  <div
    className="row"
    id={undefined}
  >
    <div
      className="col-md-4 col-xs-4"
      id={undefined}
    >
      <label
        htmlFor="Incentives"
      >
        Incentives
      </label>
    </div>
    <div
      className="col-md-5 col-xs-5"
      id={undefined}
    >
      <div
        className="form-group undefined"
      >
        <div
          className="input-group"
        >
          <span
            className="input-group-addon"
          >
            <i
              className="fa fa-usd"
            />
          </span>
          <input
            className="form-control"
            disabled={true}
            id="Incentives-Input"
            name="Incentives-Input"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="0.00"
            type="text"
            value="6465"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
    id={undefined}
  >
    <div
      className="col-md-4 col-xs-4"
      id={undefined}
    >
      <label
        htmlFor="LoyaltyCash"
      >
        Loyalty Cash
      </label>
    </div>
    <div
      className="col-md-5 col-xs-5"
      id={undefined}
    >
      <div
        className="form-group undefined"
      >
        <div
          className="input-group"
        >
          <span
            className="input-group-addon"
          >
            <i
              className="fa fa-usd"
            />
          </span>
          <input
            className="form-control"
            disabled={true}
            id="LoyaltyCash-Input"
            name="LoyaltyCash-Input"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="0.00"
            type="text"
            value="565"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
    id={undefined}
  >
    <div
      className="col-md-4 col-xs-4"
      id={undefined}
    >
      <label
        htmlFor="dealerRebate"
      >
        Dealer Rebate
      </label>
    </div>
    <div
      className="col-md-5 col-xs-5"
      id={undefined}
    >
      <div
        className="row"
        id={undefined}
      >
        
        <div
          className="col-md-12 col-xs-12"
          id={undefined}
        >
          <div
            className="form-group undefined"
          >
            <div
              className="input-group"
            >
              <span
                className="input-group-addon"
              >
                <i
                  className="fa fa-usd"
                />
              </span>
              <input
                className="form-control"
                disabled={false}
                id="DealerRebate-1-Input"
                name="DealerRebate-1-Input"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="0.00"
                type="text"
                value="5454"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="row"
        id={undefined}
      >
        
        <div
          className="col-md-12 col-xs-12"
          id={undefined}
        >
          <div
            className="form-group undefined"
          >
            <div
              className="input-group"
            >
              <span
                className="input-group-addon"
              >
                <i
                  className="fa fa-usd"
                />
              </span>
              <input
                className="form-control"
                disabled={false}
                id="DealerRebate-2-Input"
                name="DealerRebate-2-Input"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="0.00"
                type="text"
                value="321"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="row"
        id={undefined}
      >
        
        <div
          className="col-md-12 col-xs-12"
          id={undefined}
        >
          <div
            className="form-group undefined"
          >
            <div
              className="input-group"
            >
              <span
                className="input-group-addon"
              >
                <i
                  className="fa fa-usd"
                />
              </span>
              <input
                className="form-control"
                disabled={false}
                id="DealerRebate-3-Input"
                name="DealerRebate-3-Input"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="0.00"
                type="text"
                value="33"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-3 col-xs-3"
      id={undefined}
    >
      <div
        className="extra-padded-row row"
        id={undefined}
      >
        <div
          className="dealer-rebate-add col-md-6 col-xs-6"
          id={undefined}
        >
          <a
            className="disabled"
            disabled={true}
            id="dealer-rebate-add-1"
            onClick={false}
          >
            <i
              className="fa fa-plus"
            />
          </a>
        </div>
        <div
          className="dealer-rebate-remove col-md-6 col-xs-6"
          id={undefined}
        >
          <a
            className=""
            disabled={false}
            id="dealer-rebate-remove-1"
            onClick={[Function]}
          >
            <i
              className="fa fa-minus"
            />
          </a>
        </div>
      </div>
      <div
        className="extra-padded-row row"
        id={undefined}
      >
        <div
          className="dealer-rebate-add col-md-6 col-xs-6"
          id={undefined}
        >
          <a
            className="disabled"
            disabled={true}
            id="dealer-rebate-add-2"
            onClick={false}
          >
            <i
              className="fa fa-plus"
            />
          </a>
        </div>
        <div
          className="dealer-rebate-remove col-md-6 col-xs-6"
          id={undefined}
        >
          <a
            className=""
            disabled={false}
            id="dealer-rebate-remove-2"
            onClick={[Function]}
          >
            <i
              className="fa fa-minus"
            />
          </a>
        </div>
      </div>
      <div
        className="extra-padded-row row"
        id={undefined}
      >
        <div
          className="dealer-rebate-add col-md-6 col-xs-6"
          id={undefined}
        >
          <a
            className=""
            disabled={false}
            id="dealer-rebate-add-3"
            onClick={[Function]}
          >
            <i
              className="fa fa-plus"
            />
          </a>
        </div>
        <div
          className="dealer-rebate-remove col-md-6 col-xs-6"
          id={undefined}
        >
          <a
            className=""
            disabled={false}
            id="dealer-rebate-remove-3"
            onClick={[Function]}
          >
            <i
              className="fa fa-minus"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
